### PR TITLE
lapack/{gonum,testlapack}: clean up Dorgr2 and its test

### PR DESCRIPTION
**Pull request by vladimir-ch**
PTAL

cc @soypat

### CODE REVIEW

File: lapack/gonum/dorgr2.go [hunk 1]
```diff
@@ -54,16 +49,16 @@ func (impl Implementation) Dorgr2(m, n, k int, a []float64, lda int, tau, work [
...
-		for l := 0; l < m-k; l++ {
-			for j := 0; j < n; j++ {
-				a[l*lda+j] = 0
-			}
-			a[l*lda+n-m+l] = 1
+	// Initialise rows 0:m-k to rows of the unit matrix.
+	for l := 0; l < m-k; l++ {
+		for j := 0; j < n; j++ {
+			a[l*lda+j] = 0
 		}
```
Comment (kortschak) [inline lapack/gonum/dorgr2.go:60]: Is it worth helping the compiler optimise this?
```
row := a[l*lda:l*lda+n]
for j := range row {
    row[j] = 0
}
```

File: lapack/testlapack/dorgr2.go [hunk 2]
```diff
@@ -35,60 +37,72 @@ func Dorgr2Test(t *testing.T, impl Dorgr2er) {
...
-		}
-	}
 	// Compute the matrix Q using Dorg2r.
-	impl.Dorgr2(m, n, k, a.Data, a.Stride, tau[m-k:m], work)
+	q := cloneGeneral(rq)
+	impl.Dorgr2(m, n, k, q.Data, q.Stride, tau[m-k:m], work)
+
 	if m == 0 {
 		return
 	}
```
Comment (kortschak) [inline lapack/testlapack/dorgr2.go:64]: Does this happen this late for panic checks?

Comment (vladimir-ch) [reply lapack/testlapack/dorgr2.go:64]: Yes, to check for panics but also for modifications if there are non-zero-sized const input slices.

File: lapack/gonum/dorgr2.go (hunk 1 above)
Comment (vladimir-ch) [reply lapack/gonum/dorgr2.go:60]: Good idea, done.

File: lapack/testlapack/dorgr2.go (hunk 2 above)
Comment (kortschak) [reply lapack/testlapack/dorgr2.go:64]: > but also for modifications if there are non-zero-sized const input slices.

I'm confused about this. There are no test checks prior to this return, so if m is zero, the only thing that can be observed as a failure is a panic. What am I missing?

Comment (vladimir-ch) [reply lapack/testlapack/dorgr2.go:64]: Ah, sorry for not being clear, I meant this comment in general. You're right that here I postpone the quick return only for a panic check. In other routines there might also be returned errors or inputs that I check for modifications between the call and the quick return (for example, Dtbtrs).

Comment (kortschak) [reply lapack/testlapack/dorgr2.go:64]: OK. Thanks.

**Comment by soypat** [comment]
Much better. LGTM. I'll definitely carry over some style decisions to future PRs. Thanks Vladimir

**Review by soypat — APPROVED** [review]
Much better. LGTM. I'll definitely carry over some style decisions to future PRs. Thanks Vladimir

**Comment by vladimir-ch** [comment]
@kortschak Can you please approve too?

**Review by kortschak — APPROVED** [review]

**Pull request merged**